sdk/core/Azure.Core/src/Shared/ConnectionString.cs

```diff
@@ -0,0 +1,248 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Azure.Core
+{
+    internal sealed class ConnectionString
+    {
+        private readonly Dictionary<string, string> _segments;
+        private readonly string _segmentSeparator;
+        private readonly string _keyValueSeparator;
+
+        public static ConnectionString Parse(string connectionString, string segmentSeparator = ";", string keyValueSeparator = "=", bool allowFlags = false, bool allowEmptyValues = false, bool allowWhitespaces = false)
+        {
+            Validate(connectionString, segmentSeparator, keyValueSeparator, allowFlags, allowEmptyValues, allowWhitespaces);
+            return new ConnectionString(ParseSegments(connectionString, segmentSeparator, keyValueSeparator), segmentSeparator, keyValueSeparator);
+        }
+
+        private ConnectionString(Dictionary<string, string> pairs, string segmentSeparator, string keyValueSeparator)
+        {
+            _segments = pairs;
+            _segmentSeparator = segmentSeparator;
+            _keyValueSeparator = keyValueSeparator;
+        }
+
+        public string GetRequired(string key)
```

Comment:
I'm a little worried about introducing terms like "key" and "flag" here (which also show up in the test code), even in an internal type, when these terms are so overloaded.  At a minimum, we should have comments defining them for folks coming to the code for the first time, or folks coming back to the code after they've been away for a while.  But ideally, we'd use more descriptive names.  Segment and pairs I'm fine with.

Reply:
Renamed to `keyword`